README.md
# GChandler

GChandler = Garmin Connect (GG) + handler

If you are a GC user, with `gchandler` you can programmatically login into GC with your username an password and then handle it. That is, you can see  your userinfo, reveals your  list of activities, downoads the activity, etc. If it is known a GC endpoint with infornation contained, a new item can be added into the index page.

## Motivation

Ι hate the official GC for tons of unnecessary information, lack of the needed data thought this data is inside FIT files, slowness, ugliness, etc. And I am still learning JS.

## Installation

### Installation for non-programmers who found this program in github

- Click on the green button, download everything as zip file and unzip.
- Install [nodejs](https://nodejs.org/en/)
- Install [fitplotter](https://github.com/karaul/fitplotter/)  and [fitalyser](https://github.com/karaul/fitalyser/), so the whole set must be as
-The working installation is:
 ```
/GChandler
/fitalyser
/fitplotter
```
- click `gchandler.bat` -  the start page will be opened
- login to Garmin Connect with your credentials
- play with buttons
- to be continued

### Installation for programmers who found it in github

- to be continued

## Usage

- to be continued

Read [CHANGELOG.md](https://github.com/karaul/gchandler/blob/master/CHANGELOG.md) for last changes - there are   explanations how to use the program too.

## How it works

`gchandler` needs a static http-server (to be continued)

## Acknowledgments

GC login follows the logics of  [garmin-uploader](https://github.com/La0/garmin-uploader/blob/master/garmin_uploader/api.py) written by Bastien Abadie;

## License

MIT license

(c) Evgeny Votyakov aka karaul ([about me](http://www.irc-club.ru/karaul.html) in Russian), Nicosia, Cyprus, 2021


CHANGELOG.md
# Changelog

`GChandler` changes are documented in this file

## To do

- many things

### 2021-03-21

- published in open access

### 2021-03-14

- Τhe project started on the Pi day


src/gc.js
'use strict';

const axios = require('axios').default;
axios.defaults.withCredentials = true;

const qs = require('qs');
const fs = require('fs');
const stream = require('stream');
const path = require('path');
const util = require('util');
const pipeline = util.promisify(stream.pipeline);

const admzip = require('adm-zip');

let config;
try {
  const m = require("./.gc-config.js");
  config = m.config;
} catch (error) {
  config = { dowloadDir: "./"};
}

// https://stackoverflow.com/questions/63291156/download-zip-with-axios-and-unzip-with-adm-zip-in-memory-electron-app

// https://github.com/La0/garmin-uploader/blob/master/garmin_uploader/api.py
// https://stackoverflow.com/questions/66010962/javascript-equivalent-of-python-requests-session
// https://stackoverflow.com/questions/52549079/does-axios-support-set-cookie-is-it-possible-to-authenticate-through-axios-http

/*
const GC_MODERN = 'https://connect.garmin.com/modern';
const BASE_URL = `${GC_MODERN}/proxy`;
const DOWNLOAD_SERVICE = `${BASE_URL}/download-service`;

const ACTIVITY_SERVICE = `${BASE_URL}/activity-service`;
const ACTIVITYLIST_SERVICE = `${BASE_URL}/activitylist-service`;
const CURRENT_USER_SERVICE = `${GC_MODERN}/currentuser-service/user/info`;
const DEVICE_SERVICE = `${BASE_URL}/device-service`;
*/

const
    URL_HOSTNAME = 'https://connect.garmin.com/modern/auth/hostname',
    URL_LOGIN = 'https://sso.garmin.com/sso/login',
    URL_POST_LOGIN = 'https://connect.garmin.com/modern/',
    URL_PROFILE = 'https://connect.garmin.com/modern/currentuser-service/user/info',
    URL_HOST_SSO = 'sso.garmin.com',
    URL_HOST_CONNECT = 'connect.garmin.com',
    URL_SSO_SIGNIN = 'https://sso.garmin.com/sso/signin',
    URL_ACTIVITIES_LIST = "https://connect.garmin.com/proxy/activitylist-service/activities/search/activities",
    URL_UPLOAD = 'https://connect.garmin.com/modern/proxy/upload-service/upload',
    URL_ACTIVITY_TYPES = 'https://connect.garmin.com/proxy/activity-service/activity/activityTypes';

const
    URL_ACTIVITY_HEAD = (id) => `https://connect.garmin.com/proxy/activity-service/activity/${id}/details`,
    URL_ACTIVITY_BODY = (id) => `https://connect.garmin.com/proxy/download-service/files/activity/${id}`;


// Who needs mandatory urls in a request parameters !
/*var params = {
    'gauthHost': 'https://sso.garmin.com/sso',
    'service': 'https://connect.garmin.com/modern/',
}*/

var params = {
    'clientId': 'GarminConnect',
    'connectLegalTerms': 'true',
    'consumeServiceTicket': 'false',
    'createAccountShown': 'true',
    'cssUrl': 'https://connect.garmin.com/gauth-custom-v1.2-min.css',
    'displayNameShown': 'false',
    'embedWidget': 'false',
    'gauthHost': 'https://sso.garmin.com/sso',
    'generateExtraServiceTicket': 'true',
    'generateNoServiceTicket': 'false',
    'generateTwoExtraServiceTickets': 'true',
    'globalOptInChecked': 'false',
    'globalOptInShown': 'true',
    'id': 'gauth-widget',
    'initialFocus': 'true',
    'locale': 'fr_FR',
    'locationPromptShown': 'true',
    'mfaRequired': 'false',
    'mobile': 'false',
    'openCreateAccount': 'false',
    'privacyStatementUrl': 'https://www.garmin.com/fr-FR/privacy/connect/',
    'redirectAfterAccountCreationUrl': 'https://connect.garmin.com/modern/',
    'redirectAfterAccountLoginUrl': 'https://connect.garmin.com/modern/',
    'rememberMeChecked': 'false',
    'rememberMeShown': 'true',
    'rememberMyBrowserChecked': 'false',
    'rememberMyBrowserShown': 'false',
    'service': 'https://connect.garmin.com/modern/',
    'showConnectLegalAge': 'false',
    'showPassword': 'true',
    'showPrivacyPolicy': 'false',
    'showTermsOfUse': 'false',
    'source': 'https://connect.garmin.com/signin/',
    'useCustomHeader': 'false',
};

var headers = {
    'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:48.0) Gecko/20100101 Firefox/50.0',
    'NK': 'NT',
    'Cookie': []
};


var common_headers = {
    'NK': 'NT',
};


function setHeadersCookie(r) {
    const setCookies = r && r.headers && r.headers['set-cookie'];
    headers.Cookie = headers.Cookie || [];
    if (setCookies) {
        headers.Cookie = headers.Cookie.concat(setCookies);
    }
}

function errorHandler(r, message) {
    console.log("My catch reports an error: " + message);
    let status = "status" in r ? r.status : 999999;
    //console.log("r:" + typeof r);
    //console.log("status: " + status);
    if (status == 429) {
        console.log("statusText: " + r.statusText);
        console.log("retry after:" + r.headers['retry-after']);
        throw new Εrror("status 429");
    }
    if (status == 500) {
        throw new Εrror("status 500");
    }
    return status;
}

async function nextget(location) { // params: params
    return await axios.get(location, {
            headers: headers,
            maxRedirects: 0
        })
        .then((r) => {
            return r;
        }).catch((r) => {
            let status = errorHandler(r.response, "nextget");
            if (status == 302 || status == 301) {
                return r.response;
            } else {
                console.log(typeof r.response);
                console.log(Object.entries(r.response));
                throw new Εrror("Error status: " + status);
            }

        })
}

// https://staxmanade.com/2016/07/easily-simulate-slow-async-calls-using-javascript-async-await/
async function stall(stallTime = 3000) {
    await new Promise(resolve => setTimeout(resolve, stallTime));
}

module.exports = {

    test: async function (data) {
        let result = "test gc, data: " + data;
        return result;
    },

    loginTest: async function (username, password, verbose) {
        if (verbose) {
            console.log("enter test login:");
            console.log("username: " + username);
            console.log("password: " + password);
        }
        //var loginStatus=  await stall(500); // stalls for 500 ms = 0.5 sec
        var loginStatus=  await stall() // stalls for 3000 ms deault
            .then( r=> { 
                let s={};
                if (username === "goodUser") {
                    s.login = "OK";
                } else {
                    s.login = "NO"
                    s.error = "not good User"
                }
                return s; 
            }) //{ return {login: "OK"} } )
            .catch( r=> { return {login: "NO", error: "catch error"} })
        
        return loginStatus;
    },

    userinfoTest: async function () {
        console.log("enter userinfoTest:");
        var userinfo =  await stall() // stalls for 3000 ms deault
            .then( r=> { return {displayName: "bestRunner"} } )
            .catch( r=> { return {displayName: "undedefined"} } )
        return userinfo;
    },

    // config contains doawnloadDir and etc
    config: config,

    login: async function (username, password, verbose) {

        let loginStatus = {login: "NO"};

        var verbose = verbose || false;
        var data = {
            'embed': 'false',
            'username': username,
            'password': password,
        }

        // get host
        let webhost = await axios.get(URL_HOSTNAME, {
                headers: headers
            })
            .then((r) => {
                return r.data.host;
            })
            .catch((r) => {
                loginStatus.error = "No host";
                return null;
                //throw Error("No host")
            })

        if ("error" in loginStatus) return loginStatus;

        params['webhost'] = webhost;
        if (verbose) console.log("host: " + webhost);

        // get csrf_token
        let csrf_token = await axios.get(URL_LOGIN, {
                headers: headers,
                params: params
            }).then((r) => {
                let t = r.data;
                let s = t.substring(t.indexOf('input type="hidden" name="_csrf"'));
                let csrf_token = s.slice(0, s.indexOf("/>")).split('"')[5] || "";
                return csrf_token;
            })
            .catch((r) => {
                loginStatus.error = "No csrf_token";
                return null;
                //throw Error("No csrf_token")
            })
        // if (!csrf_token) return loginStatus;

        if ("error" in loginStatus) return loginStatus;

        data['_csrf'] = csrf_token;
        if (verbose) console.log("csrf_token: " + csrf_token);

        headers['Host'] = URL_HOST_SSO;
        headers['Referer'] = URL_SSO_SIGNIN;

        // get ticket
        let ticket = await axios.post(URL_LOGIN, qs.stringify(data), {
                headers: headers,
                params: params
            })
            .then((r) => {
                let t = r.data;
                //console.log("passed");
                //console.log(r);
                let s = t.substring(t.indexOf('var response_url'));
                let ticket = s.slice(s.indexOf("?ticket"), s.indexOf("\";")); // +8
                //console.log("\ncookies after post\n");
                //console.log(r.headers['set-cookie']);
                setHeadersCookie(r);
                return ticket;
            })
            .catch((r) => {
                loginStatus.error = "Post error, no ticket";
                return null;
                //throw Error("Post error, no ticket");
            });

        //if (!ticket) return loginStatus;

        if ("error" in loginStatus) return loginStatus;

        if (verbose) {
            console.log(ticket);
            console.log("\nheaders.Cookie:");
            console.log(headers.Cookie);
        }

        headers = {};
        headers['Host'] = URL_HOST_CONNECT;

        let status = null;
        // auxialiry function to get final cookies
        async function getcookies() {
            let location = URL_POST_LOGIN + ticket;
            for (let k = 0; k < 8; k++) {
                console.log("k=" + k);
                status = await nextget(location).then(r => {
                    //console.log(r);
                    status = r.status;
                    location = r.headers.location;
                    console.log("status:" + status);
                    console.log("location: " + location);
                    console.log("Current set-cookie");
                    console.log(r.headers['set-cookie']);
                    setHeadersCookie(r);
                    return status;
                })
                if (status == 200) break;
            }
            if (!(status == 200)) loginStatus.error = "No proper redirection";
        }

        if ("error" in loginStatus) return loginStatus;

        // get final cookies
        let cookies = await getcookies().then(status => {
                if (verbose) {
                    console.log("getcookies status: " + status);
                    console.log("Final cookies");
                    console.log(headers.Cookie);
                }
                return headers.Cookie;
            })
            .catch(() => {
                loginStatus.error = "No final cookies";
                return null;
                //throw Error("No coockies")
            })
        //if (!cookies) return loginStatus;
        
        if ("error" in loginStatus) return loginStatus;

        loginStatus.login = "OK";
        return loginStatus;
    },


    // check login
    userinfo: async function () {
        const userinfo =  await axios.get(URL_PROFILE, //, params: params
            {
                headers: headers
            }).then((r) => {
            //console.log("\nuserinfo set-cookie\n");    
            //console.log(r.headers['set-cookie']);
            //console.log("r.data");
            //console.log(r.data);
            return r.data;
        })
        .catch(() => {
            throw Error("No userinfo")
        })
        return userinfo
    },

    // retrieve a list of the activities
    //let start_index = 0;
    //let max_limit = 4;
    activitiesList: async function(start_index, max_limit) {
        const activitiesList =  await axios.get(URL_ACTIVITIES_LIST, //, params: params
                {
                    headers: headers,
                    params: {
                        "start": start_index,
                        "limit": max_limit
                    }
                })
            .then((r) => {
                return r.data;
            })
            .catch(() => {
                throw Error("No activitiesList")
            })
        return activitiesList;
    },

    // activities id:
    //activitiesList.forEach(el => {
    //   let id = el.activityId;
    //    console.log("activityID: " + id);
    //    console.log("URL: " + URL_ACTIVITY_BODY(id))
    //});    

    // retrieve a head of the activities
    //let id = activitiesList[0].activityId;
    activitiesHead: async function (id) {
        //var id = activity.activityId || activity;
        const activitiesHead = await axios.get(URL_ACTIVITY_HEAD(id), //, params: params
                {
                    headers: headers
                })
            .then((r) => {
                return r.data;
            })
            .catch(() => {
                throw Error("No activitiesHead")
            })
        return activitiesHead;
    },

    /*
    let downloadDir = "./../myactivities/";
    let k = 0;
    let activity = activitiesList[k];
    id = activitiesList[k].activityId;
    */

    // download activity
    downloadActivity: async function downloadActivity(downloadDir, fileName, id) {
        //const id = activity.activityId;
        //const time = new Date(activity.startTimeLocal).toISOString();
        //let fileName = time.replace(/:/g, "_") + "_" + id + ".fit";
        const url = URL_ACTIVITY_BODY(id);
        let result;
        await axios.get(url, //, params: params
                {
                    headers: headers,
                    responseType: 'arraybuffer'
                })
            .then(r => {
                const zipfile = new admzip(r.data);
                const filenameOrig = zipfile.getEntries()[0].name;
                if( fileName.slice(-4) === ".fit" ) {
                    zipfile.extractAllTo(downloadDir, true);
                    fs.rename(downloadDir + "/" + filenameOrig, downloadDir + "/" + fileName, function (err) {
                        if (err) console.log('ERROR at renaming: ' + fileName + err);
                    })
                    console.log(fileName + ": OK")
                    result = fileName;
                } else{
                    // does not work as arraybuffer
                    //result = readAsText(zipfile.getEntry(filenameOrig));
                }
            })
            .catch(r => {
                console.log("Error downloadActivity: " + fileName )
            })
        return result
    }
}

src/userinfo.js
document.addEventListener('DOMContentLoaded', function () {
    
    function httpRequestOnLoad() {
        if (this.readyState === 4 && this.status === 200) {
            switch (params.foo) {
                case "userinfo":
                    var userinfo = this.response;
                    document.getElementById("userinfo").value = userinfo.displayName;
                    makeTable(userinfo);
                    break;
                default:
                    break;
            }
        }
        if (this.status === 404) {
            throw Error("Errow with httpRequestOnLoad");
        }
    }

    function formatParams(params) {
        return "?" + Object
            .keys(params)
            .map(function (key) {
                return key + "=" + encodeURIComponent(params[key])
            })
            .join("&")
    };

    var url = "/";
    var params = {
        "foo": "userinfo", //activitiesList", // "userinfo",
        "start_index": 0,
        "max_limit": 20,
        "downloadDir": "./",
        "fileName": "test",
        "id": "1234567890"
    };

    function makeXMLHttpRequest(params) {
        let xhr = new XMLHttpRequest();
        xhr.onload = httpRequestOnLoad;
        //xhr.onreadystatechange = httpRequestOnLoad;
        xhr.open('GET', url + formatParams(params), true);
        xhr.setRequestHeader('Content-type', 'application/json; charset=utf-8');
        xhr.responseType = 'json'; // for bynary 'arraybuffer'
        xhr.onerror = function (e) {
            console.log(error(xhr.statusText));
        }
        xhr.send(null);
    }

    params.foo = "userinfo";
    makeXMLHttpRequest(params);

})


function makeTable(userinfo) {


    var row, cell;
    table = document.createElement("table");
    table.Id = "mytable";
    table.style.display = "block";
    table.style.overflow = "auto";
    table.style.height = "700px";
    table.style.width = "600px";

    const tHead = table.createTHead();
    row = tHead.insertRow();
    row.style = "color: #fff; background-color: #555;";
    //row.style.position = "sticky";
    //row.style.top = "0";
    let cellstyle = "text-align:right; word-wrap:break-word; position:sticky; top:0;";
    let headers = ["key", "value"];
    headers.forEach(p => {
        cell = row.insertCell();
        cell.textContent = p;
        cell.tabIndex = 0;
        switch (p) {
            case "key":
                cell.style = "text-align:right; word-wrap:break-word; width=100px"
                break;
            case "value":
                cell.style = "text-align:left; word-wrap:break-word;  width=500px"
                break;
        }
    });
    //tHead.addEventListener("click", sortByColumn);
    //tHead.addEventListener("keyup", sortByColumn);

    //console.log(userinfo);
    const tBody = table.createTBody();
    for (let [key, value] of Object.entries(userinfo)) {
        row = tBody.insertRow();
        headers.forEach(p => {
            cell = row.insertCell();
            switch (p) {
                case "key":
                    cell.textContent = key;
                    cell.style = "text-align:right; word-wrap:break-word;"
                    break;
                case "value":
                    cell.textContent = value;
                    cell.style = "text-align:left; word-wrap:break-word; width=500px"
                    break;
            }
        });
    }
    document.body.appendChild(table);
    resizableGrid(table);
}
//----- create and fill the table ------------------

//------- make nice view of the table -------------
function resizableGrid(table) {
    var row = table.getElementsByTagName('tr')[0],
        cols = row ? row.children : undefined;
    if (!cols) return;

    table.style.overflow = 'hidden';

    var tableHeight = table.offsetHeight;

    for (var i = 0; i < cols.length; i++) {
        var div = createDiv(tableHeight);
        cols[i].appendChild(div);
        cols[i].style.position = 'relative';
        setListeners(div);
    }

    function setListeners(div) {
        var pageX, curCol, nxtCol, curColWidth, nxtColWidth;

        div.addEventListener('mousedown', function (e) {
            curCol = e.target.parentElement;
            nxtCol = curCol.nextElementSibling;
            pageX = e.pageX;

            var padding = paddingDiff(curCol);

            curColWidth = curCol.offsetWidth - padding;
            if (nxtCol)
                nxtColWidth = nxtCol.offsetWidth - padding;
        });

        div.addEventListener('mouseover', function (e) {
            e.target.style.borderRight = '2px solid #0000ff';
        })

        div.addEventListener('mouseout', function (e) {
            e.target.style.borderRight = '';
        })

        document.addEventListener('mousemove', function (e) {
            if (curCol) {
                var diffX = e.pageX - pageX;

                if (nxtCol)
                    nxtCol.style.width = (nxtColWidth - (diffX)) + 'px';

                curCol.style.width = (curColWidth + diffX) + 'px';
            }
        });

        document.addEventListener('mouseup', function (e) {
            curCol = undefined;
            nxtCol = undefined;
            pageX = undefined;
            nxtColWidth = undefined;
            curColWidth = undefined
        });
    }

    function createDiv(height) {
        var div = document.createElement('div');
        div.style.top = 0;
        div.style.right = 0;
        div.style.width = '5px';
        div.style.position = 'absolute';
        div.style.cursor = 'col-resize';
        div.style.userSelect = 'none';
        div.style.height = height + 'px';
        return div;
    }

    function paddingDiff(col) {

        if (getStyleVal(col, 'box-sizing') == 'border-box') {
            return 0;
        }

        var padLeft = getStyleVal(col, 'padding-left');
        var padRight = getStyleVal(col, 'padding-right');
        return (parseInt(padLeft) + parseInt(padRight));

    }

    function getStyleVal(elm, css) {
        return (window.getComputedStyle(elm, null).getPropertyValue(css))
    }
};
//----------- make nice view of the table ------------------

src/main.js
document.addEventListener('DOMContentLoaded', function () {

    var config;
    var activitiesList = {};

    window.onclick = function (event) {
        var m = document.getElementById('modal');
        if (event.target == m) {
            m.style.display = "none";
        }
    }

    document.getElementById("Login").onclick = function (e) {
        if (loginCounter == 0) {
            setTimeout(makeXMLHttpRequest({
                foo: "status"
            }), 3000);
        } else {
            var data = {
                "foo": "login", //activitiesList", // "userinfo",
                "username": document.getElementById("username").value,
                "password": document.getElementById("password").value,
            };
            var xhr = new XMLHttpRequest();
            xhr.open("POST", "/", true);
            xhr.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
            xhr.onreadystatechange = httpRequestOnLoad;
            xhr.send(formatParams(data));
        };
    }


    function pending() {
        if (!(document.getElementById("loginStatus").value === "logged")) {
            loginCounter++;
            document.getElementById("loginStatus").value = "pending  " + loginCounter + "...";
            params.foo = "status";
            makeXMLHttpRequest(params);
        }
    }

    var windowActivitiesList = null;
    document.getElementById("activitiesList").onclick = function (e) {
        if (document.getElementById("loginStatus").value === "logged") {
            if (windowActivitiesList == null || windowActivitiesList.closed) {
                let url = "LevelUpfitalyser/index.html" + "?action=GC" +
                    "&start_index=" + document.getElementById('start_index').value +
                    "&max_limit=" + document.getElementById('max_limit').value +
                    "&downloadDir=" + document.getElementById('downloadDir').value;
                windowActivitiesList = window.open(url); //'/activitieslist.html'
                //windowActivitiesList = window.open('activitieslist.html'); //'/activitieslist.html'
            } else {
                windowActivitiesList.focus();
            }
        } else {
            pending();
        }
    }

    var windowUserInfo = null;
    document.getElementById("userinfo").onclick = function (e) {
        if (document.getElementById("loginStatus").value === "logged") {
            if (windowUserInfo == null || windowUserInfo.closed) {
                windowUserInfo = window.open('/userinfo.html')
            } else {
                windowUserInfo.focus();
            }
        } else {
            pending();
        }
    }

    document.getElementById("getLatestId").onclick = function (e) {
        if (document.getElementById("loginStatus").value === "logged") {
            params.foo = "activitiesList";
            params.start_index = document.getElementById('start_index').value;
            params.max_limit = document.getElementById('max_limit').value;
            makeXMLHttpRequest(params);
        } else {
            pending();
        }
    }


    document.getElementById("downloadNewActivities").onclick = function (e) {
        if (document.getElementById("loginStatus").value === "logged") {
            const maxOnlineId = document.getElementById('maxOnlineActivityId').value || 0;
            const maxLocalId = document.getElementById('maxLocalActivityId').value;
            if (maxOnlineId == 0) {
                alert("Click GREEN BUTTON signed:\n" + 
                    "Get GarminConnect Id's for start_index, max_limit")
                return null;
            }
            if (maxOnlineId <= maxLocalId) {
                alert("All activities are downloaded, local_max_Id >= start_index_Id, where\n" + 
                    "local_max_Id = the Latest downloaded activity Id\n" +
                    "start_index_Id = the given  GarminConnect start_index\n\n" +
                    "To download ALL the activities, please set:\n" + 
                    "start_index = 0, max_limit = 999999, local_max_Id = 0\n" +
                    "WARNING: it takes time")
                return null;
            }

            params.foo = "downloadActivity";
            params.downloadDir = document.getElementById("downloadDir").value
            async function getActivity(a) {
                const id = a.activityId;
                if (id > maxLocalId) {
                    const time = new Date(a.startTimeLocal).toISOString();
                    params.fileName = time.replace(/:/g, "_").replace(".000Z", "+00_00") + "_" + id + ".fit";
                    params.id = id;
                    makeXMLHttpRequest(params); //,"json",false);
                    return "downloaded";
                }
                return "existed";
            }        
            async function downloadNewActivitues() {
                let report = {downloaded: 0, existed: 0};
                for (const a of activitiesList) {
                    await getActivity(a).then( r=>{report[r]++} )
                }
                return report;
            }
            downloadNewActivitues().then( report => {
                params.foo = "maxLocalActivityId";
                params.maxLocalActivityId = maxOnlineId;
                for ( let [key, value] of Object.entries(report) ) {
                    const s = key + ":" + value.toString();
                    addParagraph(s);
                }
                makeXMLHttpRequest(params);
                document.getElementById('maxLocalActivityId').value = maxOnlineId;
            })
        } else {
            pending();
        }
    }

    function httpRequestOnLoad() {
        if (this.readyState === 4 && this.status === 200) {
            //console.log(this)
            switch (params.foo) {
                case "maxLocalActivityId":
                    addParagraph(this.response.text);
                    break;
                case "login":
                    params.foo = "status";
                    setTimeout(makeXMLHttpRequest(params), 3000);
                    break;
                case "status":
                    if (this.response.loginFlag) {
                        if ("config" in this.response) {
                            config = this.response.config;
                            showGarminConnectControls();
                        } else {
                            let errorInfo = "Error at login:\n" + this.response.error +
                                "\nCheck username and password";
                            if (this.response.error === "No host") errorInfo += "\nRestart http-server-static"
                            alert(errorInfo);
                            document.location.reload();
                        }
                    } else {
                        loginCounter++;
                        if (loginCounter == 1) {
                            hideUsernamePassword();
                            document.getElementById("Login").dispatchEvent(new Event('click'));
                        }
                        document.getElementById("loginStatus").value = "pending  " + loginCounter + "...";
                        params.foo = "status";
                        setTimeout(makeXMLHttpRequest(params), 3000);
                    }
                    break;
                case "activitiesList":
                    if (document.getElementById("loginStatus").value === "logged") {
                        activitiesList = this.response;
                        console.log(activitiesList.slice(-1));
                        document.getElementById("maxOnlineActivityId").value = activitiesList[0].activityId;
                        document.getElementById("minOnlineActivityId").value = activitiesList.slice(-1)[0].activityId;
                        //console.log(activitiesList);
                    }
                    break;
                case "downloadActivity":
                    //let content = toArrayBuffer(this.response);
                    //console.log(this.response);
                    addParagraph("downloaded: " +
                        document.getElementById('downloadDir').value + "/" + params.fileName + " Ok");
                    break;
                default:
                    break;
            }
        }
        if (this.status === 404) {
            throw Error("Errow with httpRequestOnLoad");
        }
    }

    function addParagraph(text) {
        let p = document.createElement('p'); // is a node
        p.innerHTML = text;
        document.body.appendChild(p);
    }

    function hideUsernamePassword() {
        document.getElementById('username').style.display = "none";
        document.getElementById('password').style.display = "none";
        document.getElementById('Login').style.display = "none";
        document.getElementById('labelPassword').style.display = "none";
        document.getElementById('labelUsername').style.display = "none";
    }

    function showGarminConnectControls() {
        hideUsernamePassword();
        document.getElementById("loginStatus").value = "logged";
        document.getElementById('gcTitle').innerHTML = "Garmin Connect handler";
        document.getElementById('loginStatus').style.display = "inline";
        document.getElementById('userinfo').style.display = "inline";
        document.getElementById('start_index').style.display = "inline";
        document.getElementById('max_limit').style.display = "inline";
        document.getElementById('start_indexLabel').style.display = "inline";
        document.getElementById('max_limitLabel').style.display = "inline";
        document.getElementById('activitiesList').style.display = "inline";
        document.getElementById('downloadDir').value = config.downloadDir;
        document.getElementById('downloadDir').style.display = "inline";
        document.getElementById('downloadDirLabel').style.display = "inline";
        //document.getElementById('maxActivityIdΤitle').style.display = "inline";
        document.getElementById('maxLocalActivityId').value = config.maxLocalActivityId;
        document.getElementById('maxLocalActivityId').style.display = "inline";
        document.getElementById('maxLocalActivityIdLabel').style.display = "inline";
        document.getElementById('getLatestId').style.display = "inline";
        document.getElementById('maxOnlineActivityId').value = 0;
        document.getElementById('maxOnlineActivityId').style.display = "inline";
        document.getElementById('maxOnlineActivityIdLabel').style.display = "inline";
        document.getElementById('minOnlineActivityId').value = 0;
        document.getElementById('minOnlineActivityId').style.display = "inline";
        document.getElementById('minOnlineActivityIdLabel').style.display = "inline";
        document.getElementById('downloadNewActivities').style.display = "inline";
        //document.getElementById('activityFileNameLabel').style.display = "inline";
        //document.getElementById('activityFileName').style.display = "inline";
        //document.getElementById('activityIdLabel').style.display = "inline";
    }

    function formatParams(params) {
        return "?" + Object
            .keys(params)
            .map(function (key) {
                return key + "=" + encodeURIComponent(params[key])
            })
            .join("&")
    };

    var url = "/";
    var params = {
        "foo": "status", //activitiesList", // "userinfo",
        "start_index": 0,
        "max_limit": 10,
        "downloadDir": "./",
        "fileName": "test",
        "id": "1234567890"
    };

    function makeXMLHttpRequest(params, type = "json", asyncmode = true) {
        let xhr = new XMLHttpRequest();
        xhr.onload = httpRequestOnLoad;
        //xhr.onreadystatechange = httpRequestOnLoad;        
        xhr.open('GET', url + formatParams(params), asyncmode);
        if (type === "json") {
            xhr.setRequestHeader('Content-type', 'application/json; charset=utf-8');
            if (asyncmode) xhr.responseType = 'json';
        }
        if (type === "arraybuffer") {
            //console.log(type);
            xhr.setRequestHeader('Content-type', 'application/octet-stream');
            if (asyncmode) xhr.responseType = 'arraybuffer';
        }
        xhr.onerror = function (e) {
            console.log(error(xhr.statusText));
        }
        xhr.send(null);
    }

    var loginCounter = 0;

})